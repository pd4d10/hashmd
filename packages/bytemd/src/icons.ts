// DO NOT EDIT, generated by scripts/icon.js
export const icons = {
  heading:
    '<svg width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M3 1v13.33M12.997 1v13.33M3 7.665h9.998" stroke="currentColor" stroke-width="1.33" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  h1:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M6 8v32M25 8v32M6 24h19M34.226 24L39 19.017V40" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  h2:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M6 8v32M24 8v32M7 24h16M32 25c0-3.167 2.667-5 5-5s5 1.833 5 5c0 5.7-10 9.933-10 15h10" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  bold:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path clip-rule="evenodd" d="M24 24c5.506 0 9.969-4.477 9.969-10S29.506 4 24 4H11v20h13zM28.031 44C33.537 44 38 39.523 38 34s-4.463-10-9.969-10H11v20h17.031z" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  italic:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M20 6h16M12 42h16M29 5.952L19 42" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  quote:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M18.853 9.116c-7.53 4.836-11.714 10.465-12.55 16.887C5 36 13.94 40.893 18.47 36.497 23 32.1 20.285 26.52 17.005 24.994c-3.28-1.525-5.286-.994-4.936-3.032.35-2.039 5.016-7.69 9.116-10.323a.749.749 0 00.114-1.02L20.285 9.3c-.44-.572-.862-.55-1.432-.185zM38.679 9.116c-7.53 4.836-11.714 10.465-12.55 16.887-1.303 9.997 7.637 14.89 12.167 10.494 4.53-4.397 1.815-9.977-1.466-11.503-3.28-1.525-5.286-.994-4.936-3.032.35-2.039 5.017-7.69 9.117-10.323a.749.749 0 00.113-1.02L40.11 9.3c-.44-.572-.862-.55-1.431-.185z" fill="currentColor"/></svg>',
  link:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M26.24 16.373l-9.14-9.14c-2.661-2.661-7.035-2.603-9.769.131-2.733 2.734-2.792 7.107-.13 9.768l7.935 7.936M32.903 23.003l7.935 7.936c2.661 2.66 2.603 7.034-.13 9.768-2.735 2.734-7.108 2.792-9.77.131l-9.14-9.14M26.11 26.142c2.733-2.734 2.791-7.108.13-9.769M21.799 21.799c-2.734 2.733-2.793 7.107-.131 9.768" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  image:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" xmlns="http://www.w3.org/2000/svg"><path fill="#fff" fill-opacity=".01" fill-rule="evenodd" d="M0 0h48v48H0z"/><g transform="translate(5 8)" stroke-width="4" stroke-linejoin="round" stroke="currentColor" fill="none"><path d="M2 0h34a2 2 0 012 2v28a2 2 0 01-2 2H2a2 2 0 01-2-2V2a2 2 0 012-2z" stroke-linecap="round"/><circle stroke-linecap="round" cx="9.5" cy="8.5" r="1.5"/><path d="M10 16l5 4 6-7 17 13v4a2 2 0 01-2 2H2a2 2 0 01-2-2v-4l10-10z"/></g></svg>',
  code:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M16 13L4 25.432 16 37M32 13l12 12.432L32 37" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/><path d="M28 4l-7 40" stroke="currentColor" stroke-width="4" stroke-linecap="round"/></svg>',
  codeBlock:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M16 4c-2 0-5 1-5 5v9c0 3-5 5-5 5s5 2 5 5v11c0 4 3 5 5 5M32 4c2 0 5 1 5 5v9c0 3 5 5 5 5s-5 2-5 5v11c0 4-3 5-5 5" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  ol:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M9 4v9M12 13H6M12 27H6M6 20s3-3 5 0-5 7-5 7M6 34.5s2-3 5-1 0 4.5 0 4.5 3 2.5 0 4.5-5-1-5-1M11 38H9M9 4L6 6M21 24h22M21 38h22M21 10h22" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  ul:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M9 42a4 4 0 100-8 4 4 0 000 8zM9 14a4 4 0 100-8 4 4 0 000 8zM9 28a4 4 0 100-8 4 4 0 000 8z" stroke="currentColor" stroke-width="4" stroke-linejoin="round"/><path d="M21 24h22M21 38h22M21 10h22" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  hr:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M5 24h38M21 38h6M37 38h6M21 10h6M5 38h6M5 10h6M37 10h6" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  info:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill="#fff" fill-opacity=".01" d="M0 0h48v48H0z"/><path d="M24 44a19.937 19.937 0 0014.142-5.858A19.937 19.937 0 0044 24a19.938 19.938 0 00-5.858-14.142A19.937 19.937 0 0024 4 19.938 19.938 0 009.858 9.858 19.938 19.938 0 004 24a19.937 19.937 0 005.858 14.142A19.938 19.938 0 0024 44z" stroke="currentColor" stroke-width="4" stroke-linejoin="round"/><path fill-rule="evenodd" clip-rule="evenodd" d="M24 11a2.5 2.5 0 110 5 2.5 2.5 0 010-5z" fill="currentColor"/><path d="M24.5 34V20h-2M21 34h7" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  fullscreenOn:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M33 6h9v9M42 33v9h-9M15 42H6v-9M6 15V6h9" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  fullscreenOff:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M33 6v9h9M15 6v9H6M15 42v-9H6M33 42v-9h8.9" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  help:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M39 6H9a3 3 0 00-3 3v30a3 3 0 003 3h30a3 3 0 003-3V9a3 3 0 00-3-3z" stroke="currentColor" stroke-width="4" stroke-linejoin="round"/><path d="M24 28.625v-4a6 6 0 10-6-6" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/><path fill-rule="evenodd" clip-rule="evenodd" d="M24 37.625a2.5 2.5 0 100-5 2.5 2.5 0 000 5z" fill="currentColor"/></svg>',
  toc:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M39 6H9a3 3 0 00-3 3v30a3 3 0 003 3h30a3 3 0 003-3V9a3 3 0 00-3-3z" stroke="currentColor" stroke-width="4" stroke-linejoin="round"/><path d="M26 24H14M34 15H14M32 33H14" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  close:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill="#fff" fill-opacity=".01" d="M0 0h48v48H0z"/><path d="M8 8l32 32M8 40L40 8" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  left:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><rect x="6" y="6" width="28" height="36" rx="2" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/><path d="M42 6v36" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  right:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><rect x="14" y="6" width="28" height="36" rx="2" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/><path d="M6 6v36" stroke="currentColor" stroke-width="4" stroke-linecap="round" stroke-linejoin="round"/></svg>',
  more:
    '<svg width="1em" height="1em" viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill="#fff" fill-opacity=".01" d="M0 0h48v48H0z"/><circle cx="12" cy="24" r="3" fill="currentColor"/><circle cx="24" cy="24" r="3" fill="currentColor"/><circle cx="36" cy="24" r="3" fill="currentColor"/></svg>',
};
