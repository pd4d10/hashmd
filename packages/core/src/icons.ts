// DO NOT EDIT, generated by "pnpm icon" command
export const icons = {
  close:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="m8 8 32 32M8 40 40 8"/></svg>',
  heading:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M12 5v38M36 5v38M12 24h24"/></svg>',
  bold: '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M24 24c5.506 0 9.969-4.477 9.969-10S29.506 4 24 4H11v20h13ZM28.031 44C33.537 44 38 39.523 38 34s-4.463-10-9.969-10H11v20h17.031Z" clip-rule="evenodd"/></svg>',
  italic:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M20 6h16M12 42h16M29 5.952 19 42"/></svg>',
  quote:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path fill="currentColor" fill-rule="evenodd" d="M18.853 9.116C11.323 13.952 7.14 19.58 6.303 26.003 5 36 13.94 40.893 18.47 36.497 23 32.1 20.285 26.52 17.005 24.994c-3.28-1.525-5.286-.994-4.936-3.033.35-2.038 5.016-7.69 9.116-10.322a.749.749 0 0 0 .114-1.02L20.285 9.3c-.44-.572-.862-.55-1.432-.185ZM38.679 9.116c-7.53 4.836-11.714 10.465-12.55 16.887-1.303 9.997 7.637 14.89 12.167 10.494 4.53-4.397 1.815-9.977-1.466-11.503-3.28-1.525-5.286-.994-4.936-3.033.35-2.038 5.017-7.69 9.117-10.322a.749.749 0 0 0 .113-1.02L40.11 9.3c-.44-.572-.862-.55-1.431-.185Z" clip-rule="evenodd"/></svg>',
  link: '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="m26.24 16.373-9.14-9.14c-2.661-2.661-7.035-2.603-9.768.131-2.734 2.734-2.793 7.107-.131 9.768l7.935 7.936M32.903 23.003l7.935 7.935c2.661 2.662 2.603 7.035-.13 9.769-2.735 2.734-7.108 2.792-9.77.13l-9.14-9.14"/><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M26.11 26.142c2.733-2.734 2.792-7.108.13-9.769M21.799 21.798c-2.734 2.734-2.792 7.108-.131 9.769"/></svg>',
  image:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M5 10a2 2 0 0 1 2-2h34a2 2 0 0 1 2 2v28a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V10Z" clip-rule="evenodd"/><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M14.5 18a1.5 1.5 0 1 0 0-3 1.5 1.5 0 0 0 0 3Z" clip-rule="evenodd"/><path stroke="currentColor" stroke-linejoin="round" stroke-width="4" d="m15 24 5 4 6-7 17 13v4a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2v-4l10-10Z"/></svg>',
  code: '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M16 13 4 25.432 16 37M32 13l12 12.432L32 37"/><path stroke="currentColor" stroke-linecap="round" stroke-width="4" d="m28 4-7 40"/></svg>',
  codeBlock:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M16 4c-2 0-5 1-5 5v9c0 3-5 5-5 5s5 2 5 5v11c0 4 3 5 5 5M32 4c2 0 5 1 5 5v9c0 3 5 5 5 5s-5 2-5 5v11c0 4-3 5-5 5"/></svg>',
  ul: '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linejoin="round" stroke-width="4" d="M9 42a4 4 0 1 0 0-8 4 4 0 0 0 0 8ZM9 14a4 4 0 1 0 0-8 4 4 0 0 0 0 8ZM9 28a4 4 0 1 0 0-8 4 4 0 0 0 0 8Z"/><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M21 24h22M21 38h22M21 10h22"/></svg>',
  ol: '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M9 4v9M12 13H6M12 27H6M6 20s3-3 5 0-5 7-5 7M6 34.5s2-3 5-1 0 4.5 0 4.5 3 2.5 0 4.5-5-1-5-1M11 38H9M9 4 6 6M21 24h22M21 38h22M21 10h22"/></svg>',
  hr: '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M5 24h38M21 38h6M37 38h6M21 10h6M5 38h6M5 10h6M37 10h6"/></svg>',
  toc: '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linejoin="round" stroke-width="4" d="M39 6H9a3 3 0 0 0-3 3v30a3 3 0 0 0 3 3h30a3 3 0 0 0 3-3V9a3 3 0 0 0-3-3Z"/><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M26 24H14M34 15H14M32 33H14"/></svg>',
  help: '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linejoin="round" stroke-width="4" d="M39 6H9a3 3 0 0 0-3 3v30a3 3 0 0 0 3 3h30a3 3 0 0 0 3-3V9a3 3 0 0 0-3-3Z"/><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M24 28.625v-4a6 6 0 1 0-6-6"/><path fill="currentColor" fill-rule="evenodd" d="M24 37.625a2.5 2.5 0 1 0 0-5 2.5 2.5 0 0 0 0 5Z" clip-rule="evenodd"/></svg>',
  write:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><rect width="28" height="36" x="6" y="6" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" rx="2"/><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M42 6v36"/></svg>',
  preview:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><rect width="28" height="36" x="14" y="6" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" rx="2"/><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M6 6v36"/></svg>',
  exitFullscreen:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M33 6v9h9M15 6v9H6M15 42v-9H6M33 42v-9h8.9"/></svg>',
  fullscreen:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M33 6h9v9M42 33v9h-9M15 42H6v-9M6 15V6h9"/></svg>',
  source:
    '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="none" viewBox="0 0 48 48"><path stroke="currentColor" stroke-linecap="round" stroke-width="4" d="M29.344 30.477c2.404-.5 4.585-1.366 6.28-2.638C38.52 25.668 40 22.314 40 19c0-2.324-.881-4.494-2.407-6.332-.85-1.024 1.636-8.667-.573-7.638-2.21 1.03-5.45 3.308-7.147 2.805A20.712 20.712 0 0 0 24 7c-1.8 0-3.532.223-5.147.634C16.505 8.232 14.259 6 12 5.03c-2.26-.97-1.026 6.934-1.697 7.765C8.84 14.605 8 16.73 8 19c0 3.314 1.79 6.668 4.686 8.84 1.93 1.446 4.348 2.368 7.054 2.822M19.74 30.662c-1.159 1.275-1.738 2.486-1.738 3.632v8.717M29.345 30.477c1.097 1.44 1.646 2.734 1.646 3.88v8.654M6 31.215c.899.11 1.566.524 2 1.24.652 1.075 3.074 5.063 5.825 5.063h4.177"/></svg>',
};
