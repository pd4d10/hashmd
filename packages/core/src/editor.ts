import { icons } from "./icons";
import type { EditorProps, Locale, EditorContext, ToolbarItem } from "./types";
import { EditorView } from "@codemirror/view";
import selectFiles from "select-files";

/**
 * Wrap text with decorators, for example:
 *
 * `text -> *text*`
 */
export function wrapText(editor: EditorView, prefix: string, suffix = prefix) {
  const selection =
    editor.state.selection.ranges.find((r) => !r.empty) ?? // find the first selection
    editor.state.wordAt(editor.state.selection.main.head) ?? // if not, try to find the word
    editor.state.selection.main;

  const { from, to } = selection;
  const text = editor.state.sliceDoc(from, to); // use from/to instead of anchor/head for reverse select

  const shouldUnwrap =
    editor.state.sliceDoc(from - prefix.length, from) === prefix &&
    editor.state.sliceDoc(to, to + suffix.length) === suffix;
  if (shouldUnwrap) {
    editor.dispatch({
      changes: {
        from: from - prefix.length,
        to: to + suffix.length,
        insert: text,
      },
      selection: {
        anchor: from - prefix.length,
        head: to - prefix.length,
      },
    });
  } else {
    editor.dispatch({
      changes: { from, to, insert: prefix + text + suffix },
      selection: {
        anchor: from + prefix.length,
        head: to + prefix.length,
      },
    });
  }
  editor.focus();
}

/**
 * replace multiple lines
 *
 * `line -> # line`
 */
export function replaceLines(
  editor: EditorView,
  replace: Parameters<Array<string>["map"]>[0],
) {
  const [selection] = editor.state.selection.ranges;
  const { from } = editor.state.doc.lineAt(selection.from);
  const { to } = editor.state.doc.lineAt(selection.to);
  const lines = editor.state.sliceDoc(from, to).split("\n");

  editor.dispatch({
    changes: { from, to, insert: lines.map(replace).join("\n") },
  });
  editor.dispatch({
    selection: {
      anchor: from,
      head: editor.state.doc.lineAt(selection.to).to, // recalculate here for updated position
    },
  });
  editor.focus();
}

/**
 * Append a block based on the cursor position
 */
export function appendBlock(
  editor: EditorView,
  content: string,
  { prefix = "", suffix = "" }: { prefix?: string; suffix?: string } = {},
) {
  prefix = "\n\n" + prefix;
  suffix = suffix + "\n";

  const end = editor.state.doc.lineAt(editor.state.selection.main.head).to;
  editor.dispatch({
    changes: { from: end, insert: prefix + content + suffix },
    selection: {
      anchor: end + prefix.length,
      head: end + prefix.length + content.length,
    },
  });
  editor.focus();
}

export function findStartIndex(num: number, nums: number[]) {
  let startIndex = nums.length - 2;
  for (let i = 0; i < nums.length; i++) {
    if (num < nums[i]) {
      startIndex = i - 1;
      break;
    }
  }
  startIndex = Math.max(startIndex, 0); // ensure >= 0
  return startIndex;
}

export const getShortcutWithPrefix = (key: string, shift = false) => {
  const shiftPrefix = shift ? "Shift-" : "";
  const CmdPrefix =
    typeof navigator !== "undefined" && /Mac/.test(navigator.platform)
      ? "Cmd-"
      : "Ctrl-";
  return shiftPrefix + CmdPrefix + key;
};

export async function handleImageUpload(
  { editor }: EditorContext,
  uploadImages: NonNullable<EditorProps["uploadImages"]>,
  files: File[],
) {
  const imgs = await uploadImages(files);
  appendBlock(
    editor,
    imgs
      .map(({ url, alt, title }, i) => {
        alt = alt ?? files[i].name;
        return `![${alt}](${url}${title ? ` "${title}"` : ""})`;
      })
      .join("\n\n"),
  );
}

export function getLeftItems(locale: Locale): ToolbarItem[] {
  return [
    {
      type: "multiple",
      title: "Headings",
      icon: icons.heading,
      actions: [1, 2, 3, 4, 5, 6].map((level) => ({
        title: locale[`h${level}` as keyof Locale],
        cheatsheet:
          level <= 3 ? `${"#".repeat(level)} ${locale.headingText}` : undefined,
        click({ detail: { editor } }) {
          replaceLines(editor, (line) => {
            line = line.trim().replace(/^#*/, "").trim();
            line = "#".repeat(level) + " " + line;
            return line;
          });
        },
      })),
    },
    {
      type: "single",
      title: locale.bold,
      icon: icons.bold,
      cheatsheet: `**${locale.boldText}**`,
      shortcut: getShortcutWithPrefix("B"),
      click({ detail: { editor } }) {
        wrapText(editor, "**");
      },
    },
    {
      type: "single",
      title: locale.italic,
      icon: icons.italic,
      cheatsheet: `*${locale.italicText}*`,
      shortcut: getShortcutWithPrefix("I"),
      click({ detail: { editor } }) {
        wrapText(editor, "*");
      },
    },
    {
      type: "single",
      title: locale.quote,
      icon: icons.quote,
      cheatsheet: `> ${locale.quotedText}`,
      click({ detail: { editor } }) {
        replaceLines(editor, (line) => "> " + line);
      },
    },
    {
      type: "single",
      title: locale.link,
      icon: icons.link,
      cheatsheet: `[${locale.linkText}](url)`,
      shortcut: getShortcutWithPrefix("K"),
      click({ detail: { editor } }) {
        wrapText(editor, "[", "](url)");
      },
    },
    {
      type: "single",
      title: locale.image,
      icon: icons.image,
      cheatsheet: `![${locale.imageAlt}](url "${locale.imageTitle}")`,
      shortcut: getShortcutWithPrefix("I", true),
      async click(ctx) {
        const fileList = await selectFiles({
          accept: "image/*",
          multiple: true,
        });

        if (fileList?.length) {
          // await handleImageUpload(
          //   ctx,
          //   uploadImages!, // FIXME:
          //   Array.from(fileList),
          // );
        }
      },
    },
    {
      type: "single",
      title: locale.code,
      icon: icons.code,
      cheatsheet: "`" + locale.codeText + "`",
      shortcut: getShortcutWithPrefix("K", true),
      click({ detail: { editor } }) {
        wrapText(editor, "`");
      },
    },
    {
      type: "single",
      title: locale.codeBlock,
      icon: icons.codeBlock,
      cheatsheet: "```" + locale.codeLang + "â†µ",
      shortcut: getShortcutWithPrefix("C", true),
      click({ detail: { editor } }) {
        appendBlock(editor, "lang", { prefix: "```", suffix: "\n```\n" });
      },
    },
    {
      type: "single",
      title: locale.ul,
      icon: icons.ul,
      cheatsheet: `- ${locale.ulItem}`,
      shortcut: getShortcutWithPrefix("U", true),
      click({ detail: { editor } }) {
        replaceLines(editor, (line) => "- " + line);
      },
    },
    {
      type: "single",
      title: locale.ol,
      icon: icons.ol,
      cheatsheet: `1. ${locale.olItem}`,
      shortcut: getShortcutWithPrefix("O", true),
      click({ detail: { editor } }) {
        replaceLines(editor, (line, i) => `${i + 1}. ${line}`);
      },
    },
  ];
}

if (import.meta.vitest) {
  const { test, expect, beforeEach, describe } = import.meta.vitest;
  let cm: EditorView;

  beforeEach(async () => {
    cm = new EditorView();
    return async () => {
      cm.destroy();
    };
  });

  describe("wrap text", () => {
    test("with selection", () => {
      cm.dispatch({ changes: { from: 0, insert: "text" } });
      cm.dispatch({ selection: { anchor: 1, head: 3 } });

      wrapText(cm, "[", "](url)");
      expect(cm.state).matchSnapshot();
      wrapText(cm, "[", "](url)");
      expect(cm.state).matchSnapshot();
    });

    test("find word", () => {
      cm.dispatch({ changes: { from: 0, insert: "text" } });

      wrapText(cm, "[", "](url)");
      expect(cm.state).matchSnapshot();
      wrapText(cm, "[", "](url)");
      expect(cm.state).matchSnapshot();
    });

    test("no word", () => {
      wrapText(cm, "[", "](url)");
      expect(cm.state).matchSnapshot();
      wrapText(cm, "[", "](url)");
      expect(cm.state).matchSnapshot();
    });

    test("with same prefix and suffix", () => {
      cm.dispatch({ changes: { from: 0, insert: "text" } });

      wrapText(cm, "*");
      expect(cm.state).matchSnapshot();
      wrapText(cm, "*");
      expect(cm.state).matchSnapshot();
    });
  });

  describe("replace lines", () => {
    test("basic", () => {
      cm.dispatch({ changes: { from: 0, insert: "line1\nline2" } });

      replaceLines(cm, (line) => "> " + line);
      expect(cm.state).matchSnapshot();
    });

    test("with selection", () => {
      cm.dispatch({ changes: { from: 0, insert: "line1\nline2" } });
      cm.dispatch({ selection: { anchor: 2, head: 8 } });

      replaceLines(cm, (line) => "> " + line);
      expect(cm.state).matchSnapshot();
    });
  });

  describe("append block", () => {
    test("basic", () => {
      cm.dispatch({ changes: { from: 0, insert: "text" } });

      appendBlock(cm, "block");
      expect(cm.state).matchSnapshot();
    });

    test("with prefix and suffix", () => {
      cm.dispatch({ changes: { from: 0, insert: "text" } });

      appendBlock(cm, "let x = 1", { prefix: "```js\n", suffix: "\n```" });
      expect(cm.state).matchSnapshot();
    });
  });
}
