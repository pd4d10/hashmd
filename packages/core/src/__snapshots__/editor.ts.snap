// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`append block > basic 1`] = `
{
  "doc": "text

block
",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 6,
        "head": 11,
      },
    ],
  },
}
`;

exports[`append block > with prefix and suffix 1`] = `
{
  "doc": "text

\`\`\`js
let x = 1
\`\`\`
",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 12,
        "head": 21,
      },
    ],
  },
}
`;

exports[`replace lines > basic 1`] = `
{
  "doc": "> line1
line2",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 0,
        "head": 7,
      },
    ],
  },
}
`;

exports[`replace lines > with selection 1`] = `
{
  "doc": "> line1
> line2",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 0,
        "head": 15,
      },
    ],
  },
}
`;

exports[`wrap text > find word 1`] = `
{
  "doc": "[text](url)",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 1,
        "head": 5,
      },
    ],
  },
}
`;

exports[`wrap text > find word 2`] = `
{
  "doc": "text",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 0,
        "head": 4,
      },
    ],
  },
}
`;

exports[`wrap text > no word 1`] = `
{
  "doc": "[](url)",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 1,
        "head": 1,
      },
    ],
  },
}
`;

exports[`wrap text > no word 2`] = `
{
  "doc": "",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 0,
        "head": 0,
      },
    ],
  },
}
`;

exports[`wrap text > with same prefix and suffix 1`] = `
{
  "doc": "*text*",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 1,
        "head": 5,
      },
    ],
  },
}
`;

exports[`wrap text > with same prefix and suffix 2`] = `
{
  "doc": "text",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 0,
        "head": 4,
      },
    ],
  },
}
`;

exports[`wrap text > with selection 1`] = `
{
  "doc": "t[ex](url)t",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 2,
        "head": 4,
      },
    ],
  },
}
`;

exports[`wrap text > with selection 2`] = `
{
  "doc": "text",
  "selection": {
    "main": 0,
    "ranges": [
      {
        "anchor": 1,
        "head": 3,
      },
    ],
  },
}
`;
